### ISSUE TITLE
Question: Is it possible to maintain the same node initial coordinates after each render?

### ISSUE THREAD

**Issue by qasimalbaqali** [vasturiano/react-force-graph#188]
tldr; is it possible to update the canvas without re-rendering the react component?

This is a use case I have been trying to achieve for the past year on and off. 

Dynamically adding new nodes will trigger a re-render and shuffling of the nodes positions. Either by updating the props, or the state of the component that ForceGraph is rendered in. 

One idea came to my mind is, if I want to keep the nodes positions the same, and when adding new nodes at later point depending on the users selected timestamp for example, I would pass all of the nodes of the final graph and in the `nodeCanvasObject` hide the ones that don't fall under that selected timestamp by passing an extra label to the `NodeObject` and choosing the same color as the background. Problem is, when toggling the flag of the nodes, as in updating the list of nodes, will also trigger the graph component to re-render since there was a change. Correct me I am wrong.

Another way that I thought that might be a solution is whenever a node is rendered it maintains it's x and y coordinates and never changes UNLESS the user moves it around, but it's position won't be effected when a new node is introduced into the graph, rather that new node will get rendered in respective to the nodes that are shown in the graph. 

Is this something that is possible to do? What would be an approach for this with the current library implementation?

Interesting read regarding this: https://www.madebymike.com.au/writing/getting-the-heck-out-of-react/

**Comment by vasturiano** [comment]
@qasimalbaqali you can always step out of React and do the DOM manipulation of a portion of the DOM tree out of React's control.

That's as simple as getting a node `ref` and modifying it to your will. You can at that point use the DOM node to instantiate the non-React library [3d-force-graph](https://github.com/vasturiano/3d-force-graph) and have it control it outside React.

That is essentially all that this React wrapper is doing actually, just exposing the inner api via React props. This is for the convenience of users familiar with the React framework.

However, I'm not certain I fully understand the motivation. Is it so that existing nodes don't move when adding new ones? That is effect is intentionally set by default and not really due to React re-rendering. The force layout engine decides to warm-up again everytime the nodes change because a new balance may need to be reached. 

But, you can actually cancel that effect, in one of two ways:
* If you don't want the force engine to re-ignite on adding nodes, so that no motion occurs (not even for the new nodes, which will be placed at the canvas center initially), you can set `cooldownTicks={0}`.
* If you want the new nodes to move but not the existing ones, you can take the approach of simply fixing the existing ones. You can do that by creating `fx`, `fy` and `fz` attributes on them (with values copied from their `x`, `y` and `z` respectively). That will tell the force engine to not affect those nodes and respect their fixed positions.

**Comment by qasimalbaqali** [comment]
@vasturiano Thank you for you very thorough and informative reply! 

A bit of a background around my use case. I have different nodes that represent different 'entities' and relationship exists between these different entities which are represented by links. As a user I want to move in time forward or backwards to see how did these relationships occur, and what created them, as in what type of actions or entities created these relationships. When moving along the timeline, this is where the problem happens, as a user I move the nodes around in order to have a better visibility on what makes sense to me (putting some nodes next to each other). But later, when I want to increase the timeline, I don't want those nodes that I already put next to each other to be reshuffled again and I have to again try and drag them around to get the positions that I placed them before.

I will go ahead with the suggestion of using the 3d-force-graph directly and leverage the `fx`,`fy`,`fz`. I was looking at the source code yesterday and noticed that they are set when using dagMode.

I want to thank you again for the massive work you are doing with this library, and how impressive this all is.
